sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,968 @@
+          throw new IOException("Unable to read data : ", e);
+        }
+        currentReader = null;
+        currentRecord = null;
+        doneReading = true;
+        return false;
+      }
+
+      @Override
+      public boolean advance() throws IOException {
+        try {
+          synchronized (currentReader) {
+            if (currentReader != null && currentReader.nextKeyValue()) {
+              currentRecord = nextPair();
+              return true;
+            }
+          }
+          currentRecord = null;
+          doneReading = true;
+        } catch (InterruptedException e) {
+          throw new IOException("Unable to read data : ", e);
+        }
+        return false;
+      }
+
+      /**
+       * Returns the pair of key and value of the next record in sequence.
+       */
+      public KV<K, V> nextPair() throws IOException, InterruptedException {
+        // Transform key if required.
+        K key =
+            transformKeyOrValue(currentReader.getCurrentKey(), keyTranslationFunction, keyCoder);
+        // Transform value if required.
+        V value = transformKeyOrValue(currentReader.getCurrentValue(), valueTranslationFunction,
+            valueCoder);
+        recordsReturned++;
+        return KV.of(key, value);
+      }
+
+      /**
+       * Returns the serialized output by cloning. Given input is transformed with the provided
+       * SimpleFunction.
+       */
+      private <T1 extends Object> T1 transformKeyOrValue(T input,
+          @Nullable SimpleFunction<T, T1> simpleFunction, Coder<T1> coder)
+          throws IOException, InterruptedException {
+        T1 output;
+        if (null != simpleFunction) {
+          output = simpleFunction.apply(input);
+        } else {
+          output = (T1) input;
+        }
+        return clone(output, coder);
+      }
+
+      /**
+       * Cloning/Serialization is required to take care of Hadoop's mutable objects if returned by
+       * RecordReader as Beam needs immutable objects.
+       */
+      private <T1 extends Object> T1 clone(T1 input, Coder<T1> coder)
+          throws IOException, InterruptedException, CoderException, ClassCastException {
+        // If the input object is not of known immutable type, clone the object.
+        if (!isKnownImmutable(input)) {
+          input = CoderUtils.clone(coder, input);
+        }
+        return input;
+      }
+
+      /**
+       * Utility method to check if the passed object is of a known immutable type.
+       */
+      private boolean isKnownImmutable(Object o) {
+        Set<Class<?>> immutableTypes = new HashSet<Class<?>>(
+            Arrays.asList(String.class, Byte.class, Short.class, Integer.class, Long.class,
+                Float.class, Double.class, Boolean.class, BigInteger.class, BigDecimal.class));
+        return immutableTypes.contains(o.getClass());
+      }
+
+      @Override
+      public KV<K, V> getCurrent() throws NoSuchElementException {
```

Comment:
Why does the nextPair method & currentRecord field need to exist at all? Why not do the work of nextPair in this method?